### PR TITLE
[LLVMCPU] Enable vectorization for pack ops.

### PR DESCRIPTION
**Pull request by hanhanW**
It introduces `SplitFullPartialTransferPass` to hook up upstream VectorTransferFullPartialRewriter pattern. The pattern splits a vector.transfer operation into an in-bounds (i.e., no out-of-bounds masking) fastpath and a slowpath. This needs to be applied before converting to LLVM dialect.

Verified that we're getting 2x-3x performance improvements from Tracy.

For LHS packing, i.e.,

```mlir
lhs_4d = iree_linalg_ext.pack %lhs
  inner_dims_pos = [0, 1]
  inner_tiles = [8, 8]
  into %lhs_4d_init
  : (tensor<256x256xi8> tensor<32x32x8x8xi8>) -> tensor<32x32x8x8xi8>
```

The performance is improved from `32.59` us to `10.01` us.

For RHS packing, i.e.,

```mlir
%rhs_4d = iree_linalg_ext.pack %rhs
  outer_dims_perm = [1, 0]
  inner_dims_pos = [1, 0]
  inner_tiles = [8, 8]
  into %rhs_4d_init
  : (tensor<256x256xi8> tensor<32x32x8x8xi8>) -> tensor<32x32x8x8xi8>
```

The performance is improved from `32.35` us to `16.11` us.

Fixes https://github.com/iree-org/iree/issues/10889

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/SplitFullPartialTransferPass.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            .Case("vector-transfers",
+                  vector::VectorTransferSplit::VectorTransfer)
+            .Default(vector::VectorTransferSplit::None));
+    patterns.add<vector::VectorTransferFullPartialRewriter>(ctx, options);
+    if (failed(applyPatternsAndFoldGreedily(getOperation(),
+                                            std::move(patterns)))) {
+      return signalPassFailure();
+    }
+  }
+};
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/SplitFullPartialTransferPass.cpp:45]: @dcaballe This is just using the upstream pattern, but I think the vector masking stuff will help this!

**Comment by hanhanW** [comment]
@chelini @nicolasvasilache FYI that we have tiling and vectorization working e2e in IREE for pack ops.

**Review by MaheshRavishankar — COMMENTED** [review]
I am trying to understand why we need the fast and slow path. I'd assume we are always on the fast path. Maybe we just need to adjust the inner tile size on x86?

**Comment by hanhanW** [comment]
> I am trying to understand why we need the fast and slow path. I'd assume we are always on the fast path. Maybe we just need to adjust the inner tile size on x86?

We're not always on fast path (i.e., unmasked case) because it can be padding mode. We'll read an incomplete tile and pad it to inner_tile sizes. 

One of the options in the pass can convert them into unmasked transfer_read (i.e., guaranteed that all the reads are in bound) with some costs.

side note that we're not able to convert masked transfer_read to LLVM dialect. We need to split the masked transfer_read/write ops for making it work e2e.

**Comment by MaheshRavishankar** [comment]
> > I am trying to understand why we need the fast and slow path. I'd assume we are always on the fast path. Maybe we just need to adjust the inner tile size on x86?
> 
> We're not always on fast path (i.e., unmasked case) because it can be padding mode. We'll read an incomplete tile and pad it to inner_tile sizes.

Is there an analysis that makes sure we dont generate the slow path if we statically know that we dont need it.... This is a branch in the innermost loop which is not great. We might need it in full generality, but maybe able to avoid it in some cases?

> 
> One of the options in the pass can convert them into unmasked transfer_read (i.e., guaranteed that all the reads are in bound) with some costs.
> 
> side note that we're not able to convert masked transfer_read to LLVM dialect. We need to split the masked transfer_read/write ops for making it work e2e.

**Comment by hanhanW** [comment]
> Is there an analysis that makes sure we dont generate the slow path if we statically know that we dont need it.... This is a branch in the innermost loop which is not great. We might need it in full generality, but maybe able to avoid it in some cases?

It's already happening. We avoid the branch when we know we don't need it. I.e., when the padding is not needed. If the padding is not needed, a transfer_read op is created; all the values of in_bounds attribute are true. The pass is a no-op in this case. See the IR before and after the pass: https://gist.githubusercontent.com/hanhanW/599a82cc70545a28026340bf34e355e8/raw

If padding is needed, the transfer_read will have at least one false value in in_bounds attribute. In this case, the pass has a couple options to handle it. The `linalg-copy` option allocates a stack buffer and copies over the data if it's incomplete tile. Otherwise it takes a subview of the complete tile. See IR:

```
// -----// IR Dump After SplitFullPartialTransfer (iree-codegen-split-full-partial-transfer) //----- //
func.func @_dynamic_pack_pad_large_dispatch_1() {
  %c231104 = arith.constant 231104 : index
  %c131072 = arith.constant 131072 : index
  %c42_i32 = arith.constant 42 : i32
  %c0 = arith.constant 0 : index
  %c1 = arith.constant 1 : index
  %c32 = arith.constant 32 : index
  %c16 = arith.constant 16 : index
  %alloca = memref.alloca() {alignment = 32 : i64} : memref<32x16xi32>
  %0 = hal.interface.constant.load[0] : i32
  %1 = hal.interface.constant.load[1] : i32
  %2 = hal.interface.constant.load[2] : i32
  %3 = hal.interface.constant.load[3] : i32
  %4 = arith.index_castui %0 : i32 to index
  %5 = arith.index_castui %1 : i32 to index
  %6 = arith.index_castui %2 : i32 to index
  %7 = arith.index_castui %3 : i32 to index
  %8 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c231104) alignment(64) : memref<?x?xi32>{%4, %5}
  memref.assume_alignment %8, 64 : memref<?x?xi32>
  %9 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c131072) alignment(64) : memref<?x?x32x16xi32>{%6, %7}
  memref.assume_alignment %9, 64 : memref<?x?x32x16xi32>
  %workgroup_id_x = hal.interface.workgroup.id[0] : index
  %workgroup_count_x = hal.interface.workgroup.count[0] : index
  %workgroup_id_y = hal.interface.workgroup.id[1] : index
  %workgroup_count_y = hal.interface.workgroup.count[1] : index
  %10 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_y]
  %11 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_y]
  scf.for %arg0 = %10 to %6 step %11 {
    %12 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg0)[%6]
    %13 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_x]
    %14 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_x]
    scf.for %arg1 = %13 to %7 step %14 {
      %15 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg1)[%7]
      %subview = memref.subview %9[%arg0, %arg1, 0, 0] [%12, %15, 32, 16] [1, 1, 1, 1] : memref<?x?x32x16xi32> to memref<?x?x32x16xi32, strided<[?, 512, 16, 1], offset: ?>>
      %16 = affine.min affine_map<(d0, d1)[s0] -> (d1 * -32 + s0, d0 * 32)>(%12, %arg0)[%4]
      %17 = affine.min affine_map<(d0, d1)[s0] -> (d1 * -16 + s0, d0 * 16)>(%15, %arg1)[%5]
      %18 = affine.apply affine_map<(d0) -> (d0 * 32)>(%arg0)
      %19 = affine.apply affine_map<(d0) -> (d0 * 16)>(%arg1)
      %subview_0 = memref.subview %8[%18, %19] [%16, %17] [1, 1] : memref<?x?xi32> to memref<?x?xi32, strided<[?, 1], offset: ?>>
      scf.for %arg2 = %c0 to %12 step %c1 {
        scf.for %arg3 = %c0 to %15 step %c1 {
          %20 = affine.apply affine_map<(d0) -> (d0 * 32)>(%arg2)
          %21 = affine.min affine_map<(d0, d1) -> (32, d0 - d1 * 32)>(%16, %arg2)
          %22 = affine.apply affine_map<(d0) -> (d0 * 16)>(%arg3)
          %23 = affine.min affine_map<(d0, d1) -> (16, d0 - d1 * 16)>(%17, %arg3)
          %subview_1 = memref.subview %subview_0[%20, %22] [%21, %23] [1, 1] : memref<?x?xi32, strided<[?, 1], offset: ?>> to memref<?x?xi32, strided<[?, 1], offset: ?>>
          %24 = arith.cmpi sge, %21, %c32 : index
          %25 = arith.cmpi sge, %23, %c16 : index
          %26 = arith.andi %24, %25 : i1
          %27:3 = scf.if %26 -> (memref<?x?xi32, strided<[?, 1], offset: ?>>, index, index) {
            scf.yield %subview_1, %c0, %c0 : memref<?x?xi32, strided<[?, 1], offset: ?>>, index, index
          } else {
            linalg.fill ins(%c42_i32 : i32) outs(%alloca : memref<32x16xi32>)
            %30 = affine.min affine_map<(d0, d1, d2) -> (d0 - d1, 32)>(%21, %c0, %c32)
            %31 = affine.min affine_map<(d0, d1, d2) -> (d0 - d1, 16)>(%23, %c0, %c16)
            %subview_2 = memref.subview %subview_1[%c0, %c0] [%30, %31] [1, 1] : memref<?x?xi32, strided<[?, 1], offset: ?>> to memref<?x?xi32, strided<[?, 1], offset: ?>>
            %subview_3 = memref.subview %alloca[0, 0] [%30, %31] [1, 1] : memref<32x16xi32> to memref<?x?xi32, strided<[16, 1]>>
            memref.copy %subview_2, %subview_3 : memref<?x?xi32, strided<[?, 1], offset: ?>> to memref<?x?xi32, strided<[16, 1]>>
            %cast = memref.cast %alloca : memref<32x16xi32> to memref<?x?xi32, strided<[?, 1], offset: ?>>
            scf.yield %cast, %c0, %c0 : memref<?x?xi32, strided<[?, 1], offset: ?>>, index, index
          }
          %28 = vector.transfer_read %27#0[%27#1, %27#2], %c42_i32 {in_bounds = [true, true]} : memref<?x?xi32, strided<[?, 1], offset: ?>>, vector<32x16xi32>
          %29 = vector.broadcast %28 : vector<32x16xi32> to vector<1x1x32x16xi32>
          vector.transfer_write %29, %subview[%arg2, %arg3, %c0, %c0] {in_bounds = [true, true, true, true]} : vector<1x1x32x16xi32>, memref<?x?x32x16xi32, strided<[?, 512, 16, 1], offset: ?>>
        }
      }
    }
  }
  return
}
```

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**